db/snapshot/backup_task.cc

```diff
@@ -72,52 +73,40 @@ future<> backup_task_impl::run(sstring data_dir) {
         //  |- <keyspace name2>
         //  |- ...
         //
-        auto ks_dir = fs::path(data_dir) / _ks;
-        auto ks_dir_lister = directory_lister(ks_dir, lister::dir_entry_types::of<directory_entry_type::directory>());
-        auto close_ks_dir_lister = deferred_close(ks_dir_lister);
-
-        snap_log.trace("Scan keyspace dir {}", ks_dir.native());
-        while (auto table_ent = ks_dir_lister.get().get()) {
-            auto snapshot_dir = ks_dir / table_ent->name / sstables::snapshots_dir / _snapshot_name;
-            auto has_snapshot = file_exists(snapshot_dir.native()).get();
+        auto snapshot_dir = fs::path(data_dir) / _ks / _cf_dir / sstables::snapshots_dir / _snapshot_name;
+        if (!file_exists(snapshot_dir.native()).get()) {
+            throw std::invalid_argument(fmt::format("snapshot does not exist at {}", snapshot_dir.native()));
+        }
+        auto snapshot_dir_lister = directory_lister(snapshot_dir, lister::dir_entry_types::of<directory_entry_type::regular>());
+        auto close_snapshot_dir_lister = deferred_close(snapshot_dir_lister);
 
-            snap_log.trace("Check table snapshot dir {} (exists={})", snapshot_dir.native(), has_snapshot);
-            if (!has_snapshot) {
-                continue;
+        while (auto component_ent = snapshot_dir_lister.get().get()) {
+            auto gh = uploads.hold();
+            auto component_name = snapshot_dir / component_ent->name;
+            auto destination = fmt::format("/{}/{}/{}", _bucket, _prefix, component_ent->name);
+            snap_log.trace("Upload {} to {}", component_name.native(), destination);
+            // Start uploading in the background. The caller waits for these fibers
+            // with the _uploads gate.
+            // Parallelizm is implicitly controlled in two ways:
+            //  - s3::client::claim_memory semaphore
+            //  - http::client::max_connections limitation
+            // FIXME -- s3::client is not abortable yet, but when it will be, need to
+            // propagate impl::_as abort requests into upload_file's fibers
+            std::ignore = _client->upload_file(component_name, destination).handle_exception([this, comp = component_name] (auto ex) {
+                snap_log.error("Error uploading {}: {}", comp.native(), ex);
+                _ex = std::move(ex);
+            }).finally([gh = std::move(gh)] {});
```

Comment:
here you move the `gh` into `finally` lambda just to keep it alive until `upload_file` finishes?

Reply:
no, i am not changing this part. do you *suggest* that i should change it?

Reply:
no, just curious what is this `move` for

Reply:
i see. so it's not related to *this* change. it's a relief =)

this is a paradigm in seastar, probably in some other chained expressions well.  the idea is to keep the `holder` alive and destroy it when the closure passed to `finally` is destroyed. the `holder` is not copyable, and is moveable. so we need to `std::move()` if when capturing it.

Reply:
I see, thanks for the clarification